documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,296 @@
+
+## Goals
+
+The goal of this library is as following:
+
+* Serve as an implementation of the IPC protocol to communicate with CoreCLR's diagnostics server.
+* Provide an easy-to-use API for any library/tools authors to utilize the IPC protocol 
+
+## Non-Goals
+
+* Provide tool-specific functionalities that are too high-level (i.e. dumping the GC heap, parsing counter payload, etc.) This will broaden the scope of this library too far and will cause complexity 
+* Parse event payloads (i.e. - This is also command-specific and can be done by other libraries.
+
+## API Descriptions
+
+At a high level, the DiagnosticsClient library provides a `CommandHandler` class for each of the command specified as part of the diagnostics IPC protocol described here: https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md#commands. For example, `EventPipeCommandHandler` handles all the `EventPipe` commands that are specified, such as `CollectTracing` and `CollectTracing2`.  For each of these commands, there should be a method that handles it. 
+
+Currently the handler classes are scattered across in different format. On top of that they are all static methods. To make it cleaner and allow a single instance of the CommandHandler class to be responsible for handling communication with a *single process*, I propose that these CommandHandler methods become instance methods and their constructor to take in the process ID of the process they are responsible for communicating with.
+
+There are also helper methods that contain various util methods for talking to the IPC channel. These are included in the `DiagnosticsIpcHelper` class.
+
+We may create additional namespaces under Microsoft.Diagnostics.Client for command-specific classes that may be useful. i.e. `Microsoft.Diagnostics.Client.EventPipe`.
+
+### VersionInfo
+This is a helper class that encodes the minimum and maximum supported versions of the diagnostics IPC protocol. This helps the developer to programmatically opt in/out of certain features easily. Currently both min/max would be 1.
+
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class VersionInfo
+    {
+        public static const int MIN_SUPPORTED_VERSION;
+        public static const int MAX_SUPPORTED_VERSION;
+    }
+}
+```
+
+### DiagnosticsCommand
+This is an enum for all the diagnostics command supported in the current version of the library.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// An enum for all the supported diagnostics IPC command
+    /// </summary>
+    public enum DiagnosticsCommand
```

Comment:
You can give an enum an explicit type, i.e., in this case `public enum DiagnosticsCommand : byte`, which will ensure it matches the size of the enum in the IPC protocol.  This would also simplify the conversions from raw number to enum to string, etc.

Additionally, is this something we want an end user to be worrying about?  Will a library author be messing with these directly or will they be exclusively using the higher-level abstractions?  If the latter, we could probably make these internal.